Add a dual handle range slider widget

nyalldawson opened the pull request:
This widget implements a slider which allows users to set a value range, with a lower and upper value.

TODO:

- [x] tests
- [x] fix hardcoded values in sizeHint()
- [x] keyboard interaction
- [x] allow dragging range from center of slider
- [x] nicely handle interactions which move the lower value over the upper or vice versa

![Peek 2020-11-24 17-29](https://user-images.githubusercontent.com/1829991/100062010-a8ee2800-2e7a-11eb-926c-ba1ed37ff81d.gif)

nyalldawson committed:
Add a dual handle range slider widget

This widget implements a slider which allows users to set a value
range, with a lower and upper value.

timlinux commented:
Nice. Could you make it so that you can e.g. right click to get spin boxes that allow you to type values in directly? I always find range sliders frustrating when I am trying to set them to a specific value.

nyalldawson commented:
@timlinux I'd just do that for each use of the widget -- I don't think it's something which should be done by the slider itself. (Eg for temporal controller the corresponding values are actually dates, so a spin box isn't appropriate)

nyalldawson committed:
Port hover/pressed state handling from QSlider

nyalldawson committed:
Nicer UX when clicking over overlapping lower/upper handles

In this case the direction of the subsequent drag dictates which slider
will be moved -- a move to the left will affect the lower part of the range,
while a move to the right will change the upper part of the range.

This allows users to set the range even when the widget initially starts
with a zero-width (or small width) range

nyalldawson committed:
Make it easier to drag lower handle exactly to upper handle value and vice versa

nyalldawson committed:
Add alternative constructor

nyalldawson committed:
Fix size hints (sync with QSlider)

nyalldawson committed:
Rename inverted appearance methods to flippedDirection in order to
clarify that they aren't quite the same as their QSlider counterparts.

nyalldawson committed:
Clamp values to preset ranges

nyalldawson committed:
Range can be dragged by clicking and dragging over the selected range

nyalldawson committed:
Use cursors to indicate interactivability

nyalldawson committed:
Shift + drag over the start or end handle will move the whole range without
changing its width

nyalldawson committed:
Fix slightly oversized selection range when in flipped direction modes

nyalldawson committed:
Add unit tests

nyalldawson committed:
Show focus rects on range slider sub components

nyalldawson committed:
Add step properties to widget

nyalldawson committed:
Working key interaction

In horizontal mode left/right,pageup/pagedown changes active slider position
while up/down flips between selecting the upper, lower or range control

In vertical mode up/down/pageup/pagedown changes the active slider position
while left/right flips between the subcontrols

nyalldawson committed:
Allow home/end presses:

If a slider is selected this will snap the slider to the start/end
of the range or the other range slider's value

If the range is selected this will move the whole selected range to
the start or end of the slider without changing its width

nyalldawson committed:
Nicer keyboard interaction when sliders are overlapping

Since it's impossible to tell which slider is focused in this case,
automatically target whichever slider makes sense given the key pressed

nyalldawson commented:
@nirvn this is ready to merge now!

nirvn reviewed (approved):
Lovely  stuff.
